src/dotnet/App.Maui/Platforms/Android/CronetMessageHandler.cs

```diff
@@ -0,0 +1,190 @@
+            .AddQuicHint("actual.chat", 443, 443)
+            .AddQuicHint("media.actual.chat", 443, 443)
+            .AddQuicHint("cdn.actual.chat", 443, 443)
+            .Build();
+    }
+
+    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+    {
+        Debug.Assert(request.RequestUri != null);
+
+        var callback = new CronetCallback(request, cancellationToken);
+        var requestBuilder = _cronetEngine.NewUrlRequestBuilder(
+                request.RequestUri.ToString(),
+                callback,
+                Executor)
+            .SetHttpMethod(request.Method.Method)
+            .SetPriority(UrlRequest.Builder.RequestPriorityMedium)
+            .DisableCache();
+
+        foreach (var header in request.Headers)
+            requestBuilder.AddHeader(header.Key, string.Join(',', header.Value));
+
+        if (request.Content != null)
+        {
+            var body = await request.Content.ReadAsByteArrayAsync(cancellationToken).ConfigureAwait(false);
+            requestBuilder.SetUploadDataProvider(UploadDataProviders.Create(body), Executor);
+            foreach (var header in request.Content.Headers)
+                requestBuilder.AddHeader(header.Key, string.Join(',', header.Value));
+        }
+        requestBuilder.Build().Start();
+
+        return await callback.Task.WaitAsync(cancellationToken).ConfigureAwait(false);
+    }
+
+    protected override void Dispose(bool disposing)
+    {
+        base.Dispose(disposing);
+        _cronetEngine.Dispose();
+    }
+
+
+    private class CronetCallback : UrlRequest.Callback
+    {
+        private const int MaxRedirectCount = 3;
+        private static readonly RecyclableMemoryStreamManager MemoryStreamManager = new ();
+
+        private readonly TaskCompletionSource<HttpResponseMessage> _result;
+
+        private int _redirectCount;
+        private MemoryStream? _responseBody;
+        private IWritableByteChannel? _responseBodyChannel;
+
+        private HttpRequestMessage RequestMessage { get; }
+        private CancellationToken CancellationToken { get; }
+
+        public Task<HttpResponseMessage> Task => _result.Task;
+
+        public CronetCallback(HttpRequestMessage requestMessage, CancellationToken cancellationToken)
+        {
+            RequestMessage = requestMessage;
+            CancellationToken = cancellationToken;
+            _result = new TaskCompletionSource<HttpResponseMessage>();
+        }
+
+        public override void OnFailed(UrlRequest p0, UrlResponseInfo p1, CronetException p2)
+        {
+            _responseBodyChannel?.DisposeSilently();
+            _responseBody?.DisposeSilently();
+            _responseBodyChannel = null;
+            _responseBody = null;
+            _result.SetException(p2);
+        }
+
+        public override void OnRedirectReceived(UrlRequest p0, UrlResponseInfo p1, string p2)
```

Comment:
why such var names?

Reply:
Is it a copy/paste?

Reply:
those are generated param names from Java classes. Should we rename them?

Reply:
IMO it's right to keep the generated names if they are the same as in Java source code.